api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupManager:
+    @dataclass
+    class GroupSizeRestriction:
+        min_number: int
+        max_number: int
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def is_acceptable(self) -> bool:
+            # Checks group size is appropriate for age range
+            restrictions = GroupManager._get_allowed_group_size(self.age_range)
+            students_num = len(self.students)
+            if restrictions.min_number > students_num:
+                return False
+            if restrictions.max_number < students_num:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {students_num} > {restrictions.max_number}"
+                )
+            return True
+
+    @staticmethod
+    def get_available_teachers() -> Iterable[Teacher]:
+        return Teacher.objects.filter(
+            status__in=(
+                TeacherStatus.AWAITING_OFFER,
+                TeacherStatus.TEACHING_ACCEPTING_MORE,
+            )
+        )
+
+    @staticmethod
+    def create_group(teacher_id: int) -> Group | None:
+        for group_candidate in GroupManager._iterate_group_candidates(teacher_id):
+            if group_candidate.is_acceptable():
+                group = Group(
+                    language_and_level=group_candidate.language_and_level,
+                    communication_language_mode=group_candidate.communication_language_mode,
+                    lesson_duration_in_minutes=DEFAULT_LESSON_DURATION_MIN,
+                    monday=group_candidate.monday,
+                    tuesday=group_candidate.tuesday,
+                    wednesday=group_candidate.wednesday,
+                    thursday=group_candidate.thursday,
+                    friday=group_candidate.friday,
+                    saturday=group_candidate.saturday,
+                    sunday=group_candidate.sunday,
+                )
+
+                group_creation_timestamp = timezone.now()
+                StatusSetter.set_status(
+                    obj=group, status=GroupStatus.PENDING, status_since=group_creation_timestamp
+                )
+                group.save()
+                group.teachers.add(group_candidate.teacher)
+                group.students.set(group_candidate.students)
+
+                # QQ: do we update teacher/student status automatically,
+                # or wait for coordinator action?
```

Comment:
Ok, great.  What kind of event the bot is triggered on though? "Offered" status sounds to me as if it's already offered, not just "ready to offer". But maybe it's fine, I don't know the intended underlying mechanic.

Reply:
Well the mechanic should be that as soon as the group is formed (by algorithm or coordinator), the back-end sends a `POST` request to a webhook in the bot and it immediately contacts the students and the teacher :) So "offered" should be fine because a coordinator/admin will get a bit more confused by a "ready to offer" instead of "offered" in the log.

Reply:
Ah great! Not a question for this PR actually, but what's the recovery mechanism going to be? If the webhook action fails (bot service is down/etc), will we know which ones to retry? That sounds like a critical path.

Reply:
That's where "ready to offer" vs "offered" could come handy. The ones that triggered 200 OK from webhook and the ones that didn't for any reason. Just thoughts, we can talk about it someplace else